### PR TITLE
ci: introduce knip for checking unused packages

### LINKED ISSUE #21948: 💬 RFC: Add knip for CI
### 🔖 Need

There are currently a lot of issues with the packages (unused/duplicate exports, unused packages etc.). 

Here is [knip](https://knip.dev/) output for the somewhat current backstage repository. Some might be false positives but they should be checked and the tool configured properly.

[knip_findings.txt](https://github.com/backstage/backstage/files/13724292/knip_findings.txt)

### 🎉 Proposal

I suggest adding [knip](https://knip.dev/) tool to be part of the CI/CD pipeline of backstage. 

### 〽️ Alternatives

_No response_

### ❌ Risks

_No response_

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

initial try with the configuration and cron job

fixes #21948

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .github/workflows/check_dependencies.yml
Comment (freben) [inline .github/workflows/check_dependencies.yml:1]: Note that all workflows of this type are named `verify_*`

File: .knip.json
Comment (freben) [inline .knip.json:19]:
```suggestion
    "plugins/*/dev/**",
    "plugins/*/src/run.ts",
    "plugins/*/src/service/standaloneServer.ts"
```

File: .knip.json
Comment (freben) [inline .knip.json:11]: Hm. What does "ignore" mean in this context? `setupTests`, `/dev/` files etc SHOULD be taken into account when deciding whether a dependency is unused or not. For example the `setupTests.ts` file is typically the only place where `@testing-library/jest-dom` is referenced and that means it should stay as a dev dependency for those packages.

But they should NOT be taken into account in the check for unused files.

Interestingly, the `scripts` folder is the cause of a lot of the dependencies in the root `package.json`, which need to stay around.

**Review by freben — COMMENTED** [review]
Alright.

Another thought: Would it be possible / make sense to treat knip just like api-reports? The current workflow updates a report continuously. But what if a per-package `knip-report.md` file was generated for every package? Possibly by the existing `api-reports` command, even. Then you'll see in pull requests that "oh look, this PR results in a new row being added to the knip-report.md, that's clearly a mistake", and we could even start rejecting additions to those files automatically in CI if we wanted to

**Comment by drodil** [comment]
> Alright.
> 
> Another thought: Would it be possible / make sense to treat knip just like api-reports? The current workflow updates a report continuously. But what if a per-package `knip-report.md` file was generated for every package? Possibly by the existing `api-reports` command, even. Then you'll see in pull requests that "oh look, this PR results in a new row being added to the knip-report.md, that's clearly a mistake", and we could even start rejecting additions to those files automatically in CI if we wanted to

Interesting idea. Would this mean that the knip would be part of `backstage-repo-tools` or a script in `scripts` or even `backstage cli`? This would indeed allow separate knip configuration files per component. Not sure about the `.md` file though, what would be the use for that? Check in CI that it's empty or?

File: .github/workflows/check_dependencies.yml
Comment (drodil) [reply .github/workflows/check_dependencies.yml:41]: Yeah of course, that's why it's TODO :)

**Comment by drodil** [comment]
I am leaving for holidays today so if someone wants to continue this work or even merge, be welcome to do so. Merry Christmas 🎄

**Comment by freben** [comment]
> Interesting idea. Would this mean that the knip would be part of `backstage-repo-tools` or a script in `scripts` or even `backstage cli`?

It might have to, then, yeah - like the api extractor is. Not sure exactly where it would fit best, but spontaneously I'd say repo-tools.

> This would indeed allow separate knip configuration files per component. Not sure about the `.md` file though, what would be the use for that? Check in CI that it's empty or?

Yeah. And to be a paper trail of the currently apparently permitted deviations from the ideal state of no warnings at all. And the key thing being, we see the output immediately on PRs even before merging them - in a world where we haven't yet made it possible to make per-package exemptions in a structured way etc.

File: .github/workflows/check_dependencies.yml
Comment (awanlin) [reply .github/workflows/check_dependencies.yml:41]: My worry is that this gets merged and we start testing it, right? then we get a bunch of noise

Comment (awanlin) [reply .github/workflows/check_dependencies.yml:41]: I created an issue: https://github.com/backstage/backstage/issues/21971

**Review by awanlin — CHANGES_REQUESTED** [review]

File: packages/app-next/knip-report.md
Comment (awanlin) [inline packages/app-next/knip-report.md:1]: There reports are awesome! When viewed here it looks nice but when you look at it outside of the PR it's all mashed together: https://github.com/drodil/backstage/blob/introduce_knip/packages/app-next/knip-report.md

Can we adjust this?

Comment (drodil) [reply packages/app-next/knip-report.md:1]: I think we need a markdown reporter for the knip, see https://github.com/webpro/knip/issues/409 .. Either included in the tool itself or then we can also build custom reporter for this like explained here https://knip.dev/features/reporters#custom-reporters

But anyways, I will push one more update for this and be gone for holidays. I think it's on ok level already and people could start working with the dependencies / configuration for knip. I think reporting format could be done in a separate PR (as well as get rid of the Vale off thing for docs quality)

Comment (awanlin) [reply packages/app-next/knip-report.md:1]: Ah, sorry, saw that link but didn't actually look at it :(

Enjoy your holiday!

**Review by freben — COMMENTED** [review]
This is looking mighty fine! @Rugvip any opinions about where the CI command is located or so?

**Comment by freben** [comment]
FYI: made a rebase to see if it goes green - remember to pull the branch to your local if you continue development

**Comment by webpro** [comment]
Just chiming in (via https://github.com/webpro/knip/issues/409) to say that this is an interesting workload for Knip. Some remarks:

- Knip v3 has this majestic behavior to totally crash (OOM) when trying to lint large projects like this. Knip v4 can handle it fine, though. You can install `knip@canary` to try it out if you like.
- Linting the full repo at once takes 130 seconds on my machine, of which a whopping 100 seconds are taken by globbing files (use `knip --performance` for metrics), so I guess plenty of potential for Knip to optimize here (I can use this repo as a great benchmark!).
- Running Knip using `--directory` is cool, lints workspaces independently. Another option is configuration like below in the root of the repo. In CI you could run the whole thing at once, and when focusing on a specific workspace there's e.g. `knip --workspace plugins/rollbar-backend`:

```json
{
  "$schema": "https://unpkg.com/knip@3/schema.json",
  "ignore": ["{packages,plugins}/*/{fixtures,migrations,templates}/**"],
  "workspaces": {
    ".": {
      "entry": ["scripts/*.{js,ts}"]
    },
    "packages/*": {
      "entry": [
        "dev/index.{ts,tsx}",
        "src/index.{ts,tsx}",
        "src/routes.ts",
        "src/run.ts"
      ],
      "jest": { "entry": "src/setupTests.ts" },
      "storybook": { "entry": "src/components/**/*.stories.tsx" },
      "ignore": [".eslintrc.js", "config.d.ts", "knexfile.js"]
    },
    "plugins/*": {
      "entry": [
        "dev/index.{ts,tsx}",
        "src/index.{ts,tsx}",
        "src/routes.ts",
        "src/run.ts"
      ],
      "jest": { "entry": "src/setupTests.ts" },
      "storybook": { "entry": "src/components/**/*.stories.tsx" },
      "ignore": [".eslintrc.js", "config.d.ts", "knexfile.js"]
    },
    "microsite": {},
    "storybook": {}
  }
}
```

- This adds `microsite` and `storybook` as extra workspaces (on top of what's in `package.json#workspaces`), since they contain a `package.json`.
- The config above is my first take at it, I don't know this project at all, I'm sure you can do better :)
- Happy to answer any questions you may have!

**Comment by freben** [comment]
@webpro Thanks for all of that insight! We're short staffed over the weekends now but we'll be sure to try out your feedback.

**Comment by drodil** [comment]
@webpro thanks a lot for this! I noticed the performance problem too when running knip from the root. And decided just to skip the configuration and go for the package specific reports as usually you don't need to run it for all packages. 

Now it should be just decide if we want to run it for the whole project or for each individual package. It would be great to be able to report also per package but run it from the root of the repo (something that might be maybe added to knip perhaps?). @freben what do you think, should we keep the per package reports or just make one big report to root?

I am still on holiday but back next week and I try to find time to make support for the markdown reports to knip. And of course finalize this if we get some conclusion.

**Comment by freben** [comment]
I think the per package break-up is pretty valuable, it makes things a fair bit more visible and actionable. But not at any cost, of course; if it's vastly faster to run it all in one go, for example, we could reconsider. Maybe a custom reporter that lets us post-split or something like that could be possible? Let's explore options

**Comment by drodil** [comment]
I tried this out with both all in one go and per package and the speed is not so much different even with canary version. And it's pretty slow as @webpro mentioned. When the v4 would be available and if any actions going on for performance?

In the other news, I implemented the markdown reporter for knip and it's here waiting for review: https://github.com/webpro/knip/pull/419

**Comment by drodil** [comment]
The markdown reporter was published for `knip` so I will update this PR to use the new reporter. 

There's also some work done for the performance here https://github.com/webpro/knip/pull/426 - @webpro do you think this helps also with the long time it currently takes to generate the reports per package?

This is a bit rebase after rebase at the moment as every time some of the plugins change, the knip report changes too so I would like to know if we are going to merge this and when to prevent unnecessary work, @freben ?

**Comment by webpro** [comment]
> Now it should be just decide if we want to run it for the whole project or for each individual package. It would be great to be able to report also per package but run it from the root of the repo (something that might be maybe added to knip perhaps?).

Note that this is possible using the `--directory` option.

This differs from the `--workspace` option, which includes ancestor and dependent workspaces.

> There's also some work done for the performance here [webpro/knip#426](https://github.com/webpro/knip/pull/426) - @webpro do you think this helps also with the long time it currently takes to generate the reports per package?

It'll especially boost full repo linting (per-workspace less so).

**Comment by drodil** [comment]
> It'll especially boost full repo linting (per-workspace less so).
@webpro Ok. Do you have anything in mind to get improvements for single directory speed? Might need some tracing to figure out what takes the time - currently it's like 5+ minutes to run this per-directory for this repository. Having those reports per directory makes sense in this project as the plugins have different owners and might want to concentrate only on their own issues. Of course, people usually run the API reports for a single or few plugins but CI runs the whole thing every time which slows down it quite a bit.

**Comment by webpro** [comment]
You can try v4 canary (but that does not have class member checks).

File: .changeset/friendly-terms-love.md
Comment (freben) [inline .changeset/friendly-terms-love.md:2]: Seems to be missing a list of packages here

**Review by freben — COMMENTED** [review]
Hi! Sorry, I rudely messed around a bit with concurrency and pushed those changes to this branch after a rebase that was also needed. I hope you don't mind.

I tried to wrap the knip report generation in a p-limit and it now does multiple directories per second on my machine, vastly speeding it up and saturating my cores. But it's still a pretty massive slowdown to what was already a slow command :/ So I guess that's still TBD - how we handle the speed issues.

Have a peek at `packages/cli/knip-report.md` - it now reports things out of `dist` which is a bit funky 🤔

**Comment by drodil** [comment]
Running these in parallel makes a lot of sense (could probably be done for other reports as well..)!

I think the problem is that `knip` does not read `.gitignore` file from the root of the repository because it's run in the package directory thus not skipping the `dist` folder. I don't know if this has also some effect on the speed as well. Another option would be to run `knip` from the root, like mentioned in the comment above, but that would make one huge report that would not be as beneficial for developers I think.

@webpro do you have any insights on this?

**Comment by webpro** [comment]
- Indeed, Knip is designed to run from the root of the repo.
- Tip: large reports could be split up according to `CODEOWNERS` (also see the `codeowners` reporter); or create a custom reporter to match your needs.

**Comment by freben** [comment]
But running from the root with `--directory` should in theory then obey gitignore right?

File: .changeset/friendly-terms-love.md
Comment (freben) [inline .changeset/friendly-terms-love.md:246]: reminder to remove everything from here except repo-tools since that's the one with actual shippable changes

**Comment by webpro** [comment]
> But running from the root with `--directory` should in theory then obey gitignore right? And still make a report for only that dir so it'd work roughly like it does right now but with better ignore handling

With Knip, using `--directory` basically equals setting the `cwd` for the process, and does not take into account or look at ancestor directories (at all).

File: packages/repo-tools/src/commands/api-reports/api-reports.ts
Comment (Rugvip) [inline packages/repo-tools/src/commands/api-reports/api-reports.ts:107]: I think we need to run this separately by default, maybe a `--knip` flag, similar to `--docs`. It's a tradeoff, but in my experience it's not super uncommon to need to do a full API reports build in different situations, and this would add quite a lot of extra time to that. Tbh given the other flags that are present it might even make sense to move this to its own command. Still like `--allow-warnings` won't apply here, and we may want to add some of our own knip specific flags. Perhaps a `backstage-repo-tools knip-reports`?

File: packages/repo-tools/src/commands/api-reports/api-extractor.ts
Comment (Rugvip) [inline packages/repo-tools/src/commands/api-reports/api-extractor.ts:1493]: With a separate flag/command we'd need to tweak this too

**Review by Rugvip — COMMENTED** [review]
Thank you for setting this up! 🙏 🚀 

I'm loving the output ✨, but it's really slow 😅. Not problematically slow, just a bit too much of an impact on the local dev experience imo.

**Review by Rugvip — CHANGES_REQUESTED** [review]

**Comment by drodil** [comment]
not stale

File: packages/repo-tools/src/commands/api-reports/api-reports.ts
Comment (drodil) [reply packages/repo-tools/src/commands/api-reports/api-reports.ts]: Moved this to own command

File: packages/cli/knip-report.md
Comment (drodil) [reply packages/cli/knip-report.md]: Introduced temporary config file for knip during generation which is cleaned after. Should get rid of the dist files as they are in ignore.

File: plugins/airbrake-backend/knip-report.md
Comment (freben) [inline plugins/airbrake-backend/knip-report.md:7]: There are a couple of false negatives I think - this one is clearly used in a test for example. Not sure why this is being missed.

**Review by freben — APPROVED** [review]
For me personally I think it's fine to merge and iterate on this one.

The persisted reports will obviously go stale quickly. Maybe we can make a scheduled github workflow that makes a fresh pr once per week or whatnot as a followup. 🤷

File: plugins/airbrake-backend/knip-report.md
Comment (drodil) [reply plugins/airbrake-backend/knip-report.md:7]: Any idea about these @webpro?

Comment (webpro) [reply plugins/airbrake-backend/knip-report.md:7]: Sorry, but I shouldn't be going into codebases for one-offs. Please open an issue in the Knip repo if you think there are false positives.

**Comment by freben** [comment]
Hey terribly sorry - this actually needs a rebase now. But then let's merge it

**Comment by drodil** [comment]
> Hey terribly sorry - this actually needs a rebase now. But then let's merge it

I will rebase and re-generate also the reports

**Review by freben — COMMENTED** [review]
Alright just one last little mistake ... (i see there are unrelated test failures, i know those are fixed in master, so let's not worry about that)

File: package.json
Comment (drodil) [reply package.json:89]: Will fix this tomorrow and rebase!

**Comment by drodil** [comment]
Fucked up the rebase somehow caused by lack of morning coffee 😆 But now it should be fine!

**Review by freben — APPROVED** [review]
<3

**Pull request merged**
